File: utils/user_info.py
import sqlite3
import pymongo
import mysql.connector
import redis
import os
from dotenv import load_dotenv

load_dotenv()

class UserInfoManager:
    def __init__(self, local_db_conn, mongo_client, mysql_conn, redis_client):
        self.local_db_conn = local_db_conn
        self.mongo_client = mongo_client
        self.mysql_conn = mysql_conn
        self.redis_client = redis_client

    def store_user_info(self, user):
        try:
            if self.local_db_conn:
                cursor = self.local_db_conn.cursor()
                cursor.execute("INSERT INTO user_info (user_id, name, avatar, display_name) VALUES (?, ?, ?, ?)",
                               (user.id, user.name, user.avatar, user.display_name))
                self.local_db_conn.commit()
        except sqlite3.Error as e:
            print(f"SQLite error while storing user info: {e}")

        try:
            if self.mongo_client:
                db = self.mongo_client[os.getenv("MONGODB_DATABASE")]
                collection = db["user_info"]
                collection.insert_one({"user_id": user.id, "name": user.name, "avatar": user.avatar, "display_name": user.display_name})
        except pymongo.errors.PyMongoError as e:
            print(f"MongoDB error while storing user info: {e}")

        try:
            if self.mysql_conn:
                cursor = self.mysql_conn.cursor()
                cursor.execute("INSERT INTO user_info (user_id, name, avatar, display_name) VALUES (%s, %s, %s, %s)",
                               (user.id, user.name, user.avatar, user.display_name))
                self.mysql_conn.commit()
        except mysql.connector.Error as e:
            print(f"MySQL error while storing user info: {e}")

        try:
            if self.redis_client:
                self.redis_client.hmset(f"user_info:{user.id}", {"name": user.name, "avatar": user.avatar, "display_name": user.display_name})
        except redis.RedisError as e:
            print(f"Redis error while storing user info: {e}")

    def monitor_user_info_changes(self, user):
        try:
            if self.local_db_conn:
                cursor = self.local_db_conn.cursor()
                cursor.execute("SELECT name, avatar, display_name FROM user_info WHERE user_id = ?", (user.id,))
                result = cursor.fetchone()
                if result:
                    old_name, old_avatar, old_display_name = result
                    if old_name != user.name or old_avatar != user.avatar or old_display_name != user.display_name:
                        self.store_user_info(user)
        except sqlite3.Error as e:
            print(f"SQLite error while monitoring user info changes: {e}")

        try:
            if self.mongo_client:
                db = self.mongo_client[os.getenv("MONGODB_DATABASE")]
                collection = db["user_info"]
                result = collection.find_one({"user_id": user.id})
                if result:
                    old_name, old_avatar, old_display_name = result["name"], result["avatar"], result["display_name"]
                    if old_name != user.name or old_avatar != user.avatar or old_display_name != user.display_name:
                        self.store_user_info(user)
        except pymongo.errors.PyMongoError as e:
            print(f"MongoDB error while monitoring user info changes: {e}")

        try:
            if self.mysql_conn:
                cursor = self.mysql_conn.cursor()
                cursor.execute("SELECT name, avatar, display_name FROM user_info WHERE user_id = %s", (user.id,))
                result = cursor.fetchone()
                if result:
                    old_name, old_avatar, old_display_name = result
                    if old_name != user.name or old_avatar != user.avatar or old_display_name != user.display_name:
                        self.store_user_info(user)
        except mysql.connector.Error as e:
            print(f"MySQL error while monitoring user info changes: {e}")

        try:
            if self.redis_client:
                result = self.redis_client.hgetall(f"user_info:{user.id}")
                if result:
                    old_name, old_avatar, old_display_name = result["name"], result["avatar"], result["display_name"]
                    if old_name != user.name or old_avatar != user.avatar or old_display_name != user.display_name:
                        self.store_user_info(user)
        except redis.RedisError as e:
            print(f"Redis error while monitoring user info changes: {e}")

File: utils/user_info.ts
import sqlite3 from 'sqlite3';
import { MongoClient } from 'mongodb';
import mysql from 'mysql2/promise';
import redis from 'redis';
import { config } from 'dotenv';

config();

class UserInfoManager {
    private localDbConn: sqlite3.Database | null;
    private mongoClient: MongoClient | null;
    private mysqlConn: mysql.Connection | null;
    private redisClient: redis.RedisClientType | null;

    constructor(localDbConn: sqlite3.Database | null, mongoClient: MongoClient | null, mysqlConn: mysql.Connection | null, redisClient: redis.RedisClientType | null) {
        this.localDbConn = localDbConn;
        this.mongoClient = mongoClient;
        this.mysqlConn = mysqlConn;
        this.redisClient = redisClient;
    }

    async storeUserInfo(user: any): Promise<void> {
        await this._storeUserInfoSqlite(user);
        await this._storeUserInfoMongodb(user);
        await this._storeUserInfoMysql(user);
        await this._storeUserInfoRedis(user);
    }

    private async _storeUserInfoSqlite(user: any): Promise<void> {
        try {
            if (this.localDbConn) {
                this.localDbConn.run("INSERT INTO user_info (user_id, name, avatar, display_name) VALUES (?, ?, ?, ?)",
                    [user.id, user.name, user.avatar, user.display_name]);
            }
        } catch (e) {
            console.error(`SQLite error while storing user info: ${e}`);
        }
    }

    private async _storeUserInfoMongodb(user: any): Promise<void> {
        try {
            if (this.mongoClient) {
                const db = this.mongoClient.db(process.env.MONGODB_DATABASE);
                const collection = db.collection("user_info");
                await collection.insertOne({ user_id: user.id, name: user.name, avatar: user.avatar, display_name: user.display_name });
            }
        } catch (e) {
            console.error(`MongoDB error while storing user info: ${e}`);
        }
    }

    private async _storeUserInfoMysql(user: any): Promise<void> {
        try {
            if (this.mysqlConn) {
                await this.mysqlConn.execute("INSERT INTO user_info (user_id, name, avatar, display_name) VALUES (?, ?, ?, ?)",
                    [user.id, user.name, user.avatar, user.display_name]);
            }
        } catch (e) {
            console.error(`MySQL error while storing user info: ${e}`);
        }
    }

    private async _storeUserInfoRedis(user: any): Promise<void> {
        try {
            if (this.redisClient) {
                await this.redisClient.hSet(`user_info:${user.id}`, { name: user.name, avatar: user.avatar, display_name: user.display_name });
            }
        } catch (e) {
            console.error(`Redis error while storing user info: ${e}`);
        }
    }

    async monitorUserInfoChanges(user: any): Promise<void> {
        await this._monitorUserInfoChangesSqlite(user);
        await this._monitorUserInfoChangesMongodb(user);
        await this._monitorUserInfoChangesMysql(user);
        await this._monitorUserInfoChangesRedis(user);
    }

    private async _monitorUserInfoChangesSqlite(user: any): Promise<void> {
        try {
            if (this.localDbConn) {
                this.localDbConn.get("SELECT name, avatar, display_name FROM user_info WHERE user_id = ?", [user.id], (err, row) => {
                    if (err) {
                        console.error(`SQLite error while monitoring user info changes: ${err}`);
                    } else if (row) {
                        const { name: oldName, avatar: oldAvatar, display_name: oldDisplayName } = row;
                        if (oldName !== user.name || oldAvatar !== user.avatar || oldDisplayName !== user.display_name) {
                            this.storeUserInfo(user);
                        }
                    }
                });
            }
        } catch (e) {
            console.error(`SQLite error while monitoring user info changes: ${e}`);
        }
    }

    private async _monitorUserInfoChangesMongodb(user: any): Promise<void> {
        try {
            if (this.mongoClient) {
                const db = this.mongoClient.db(process.env.MONGODB_DATABASE);
                const collection = db.collection("user_info");
                const result = await collection.findOne({ user_id: user.id });
                if (result) {
                    const { name: oldName, avatar: oldAvatar, display_name: oldDisplayName } = result;
                    if (oldName !== user.name || oldAvatar !== user.avatar || oldDisplayName !== user.display_name) {
                        this.storeUserInfo(user);
                    }
                }
            }
        } catch (e) {
            console.error(`MongoDB error while monitoring user info changes: ${e}`);
        }
    }

    private async _monitorUserInfoChangesMysql(user: any): Promise<void> {
        try {
            if (this.mysqlConn) {
                const [rows] = await this.mysqlConn.execute("SELECT name, avatar, display_name FROM user_info WHERE user_id = ?", [user.id]);
                if (rows.length) {
                    const { name: oldName, avatar: oldAvatar, display_name: oldDisplayName } = rows[0];
                    if (oldName !== user.name || oldAvatar !== user.avatar || oldDisplayName !== user.display_name) {
                        this.storeUserInfo(user);
                    }
                }
            }
        } catch (e) {
            console.error(`MySQL error while monitoring user info changes: ${e}`);
        }
    }

    private async _monitorUserInfoChangesRedis(user: any): Promise<void> {
        try {
            if (this.redisClient) {
                const result = await this.redisClient.hGetAll(`user_info:${user.id}`);
                if (result) {
                    const { name: oldName, avatar: oldAvatar, display_name: oldDisplayName } = result;
                    if (oldName !== user.name || oldAvatar !== user.avatar || oldDisplayName !== user.display_name) {
                        this.storeUserInfo(user);
                    }
                }
            }
        } catch (e) {
            console.error(`Redis error while monitoring user info changes: ${e}`);
        }
    }
}

export default UserInfoManager;
